compiler, ci: remove acyclic and run leak tests

alaviss opened the pull request:
## Summary
Local tests have shown that the compiler does create cycles and leak
memory during compilation. As such we are dropping all acyclic
annotations and prepare the stage to either live without them or have
them reintroduced over time.


## Details
Aside from dropping acyclic annotations, this commit prepares the
compiler for future work on memory optimizations with these changes:

- Cycle collection is disabled for the compiler itself. This is because
the compilation command is not a long-running task and we value speed
over memory usage here.

- nimsuggest is set to collect cycles only after running a command. This
should keep memory usage low during extended usage while allowing the
compilation phases to stay at top speed.

- A CI pipeline has been setup to test the compiler using LeakSanitizer,
making sure that leaks are caught before they are introduced.


Some performance numbers due to these changes:

Benchmarked via: 
`hyperfine -- "bin/nim c --forceBuild --compileOnly --verbosity:0 --hints:off --warnings:off compiler/nim.nim"`


            Time             Ratio
    Before  9.155s ± 0.044s  1.00
    After   9.599s ± 0.039s  1.05

alaviss committed:
compiler, ci: remove acyclic and run leak tests

Local tests have shown that the compiler does create cycles and leak
memory during compilation. As such we are dropping all acyclic
annotations and prepare the stage to either live without them or have
them reintroduced over time.

Aside from dropping acyclic annotations, this commit prepares the
compiler for future work on memory optimizations with these changes:

- Cycle collection is disabled for the compiler itself. This is because
  the compilation command is not a long-running task and we value speed
  over memory usage here.

- nimsuggest is set to collect cycles after running a command. This
  should keep memory usage low during extended usage while allowing the
  compilation phases to stay at top speed.

- A CI pipeline has been setup to test the compiler using LeakSanitizer,
  making sure that leaks are caught before they are introduced.

Some performance numbers due to these changes:

Tested using: hyperfine -- "bin/nim c --forceBuild --compileOnly
--verbosity:0 --hints:off --warnings:off compiler/nim.nim"

        Time             Ratio
Before  9.155s ± 0.044s  1.0
After   9.599s ± 0.039s  1.05

alaviss commented:
As an aside, here are the numbers with ORC cycle collector enabled for the compiler.

```
        Time             Ratio
Before  9.155s  ± 0.044s  1.00
After   9.599s  ± 0.039s  1.05
ORC on  11.417s ± 0.150s  1.25
```

zerbina commented on compiler/ast/ast_parsed_types.nim:
```diff
@@ -226,7 +226,7 @@ type
     literal*: string
 
-  ParsedNodeData*{.final, acyclic.} = object
+  ParsedNodeData*{.final.} = object
```
```suggestion
  ParsedNodeData*{.final, acyclic.} = object
```

zerbina commented on compiler/ast/ast_types.nim:
```diff
@@ -1638,15 +1638,15 @@ type
   PInstantiation* = ref TInstantiation
 
-  TScope* {.acyclic.} = object
+  TScope* = object
```
```suggestion
  TScope* {.acyclic.} = object
```

zerbina commented on compiler/ast/idents.nim:
```diff
@@ -22,7 +22,7 @@ import
 type
   PIdent* = ref TIdent
-  TIdent*{.acyclic.} = object
+  TIdent* = object
```
```suggestion
  TIdent*{.acyclic.} = object
```

zerbina commented on compiler/backend/cgir.nim:
```diff
@@ -169,7 +169,7 @@ type
     ## Identifies a local within a procedure.
 
-  CgNode* {.acyclic.} = ref object
+  CgNode* = ref object
```
```suggestion
  CgNode* {.acyclic.} = ref object
```

zerbina commented on compiler/sem/semdata.nim:
```diff
@@ -59,7 +59,7 @@ type
   POptionEntry* = ref TOptionEntry
   PProcCon* = ref TProcCon
-  TProcCon* {.acyclic.} = object ## procedure context; also used for top-level
+  TProcCon* = object ## procedure context; also used for top-level
```
```suggestion
  TProcCon* {.acyclic.} = object ## procedure context; also used for top-level
```

zerbina commented on compiler/sem/semtypinst.nim:
```diff
@@ -121,7 +121,7 @@ proc cacheTypeInst(c: PContext; inst: PType) =
 
 type
-  LayeredIdTable* {.acyclic.} = ref object
+  LayeredIdTable* = ref object
```
```suggestion
  LayeredIdTable* {.acyclic.} = ref object
```

zerbina commented on compiler/utils/btrees.nim:
```diff
@@ -16,7 +16,7 @@ const
 
 type
-  Node[Key, Val] {.acyclic.} = ref object
+  Node[Key, Val]  = ref object
```
```suggestion
  Node[Key, Val] {.acyclic.} = ref object
```

zerbina reviewed (changes requested):
Thank you for the PR. Having an automatic leak test for the compiler and tools is a very welcome addition, being another line of defense against bugs with destructor injection and/or the cycle collector.

As for removing `.acyclic` from all types, I strongly disagree. Most reference types used in the compiler are indeed acyclic, even though it doesn't look like it from their type definitions.

Only `TSym`, `TType`, `TNode`, `TIdObj`, and `ModuleGraph` should form cycles, all others do not, and if they actually do, then that's a bug, which ASAN would catch, that needs to be fixed.

alaviss commented:
> As for removing `.acyclic` from all types, I strongly disagree. Most reference types used in the compiler are indeed acyclic, even though it doesn't look like it from their type definitions.

While I agree that these annotations might return, I think it's best that we have this PR bring us back to a "correct" baseline, then reintroduce the annotations in later PRs.

This makes sure that we have a documented trail of progress as well as information about why those types are acyclic documented within the compiler commit history.

I'm a strong believer of make it work, make it correct then make it fast, so I'd prefer each steps to be documented.

As an aside, it seems like this change brought out some allocator bugs which caused tcompilerapi to crash, so we will have to look into that first.

zerbina commented:
> I'm a strong believer of make it work, make it correct then make it fast, so I'd prefer each steps to be documented.

But removing all `.acyclic` annotations is a step back, in that it *hides* potential issues. If, for example, `ParsedDataNode` forms a cycle, then that's flat out wrong, but it's no longer caught now. 

Also, while it's true that the other types have no dedicated commit that I know of, `CgNode` *does* have a dedicated PR that marked it as `.acyclic`.

zerbina commented:
> As an aside, it seems like this change brought out some allocator bugs which caused tcompilerapi to crash, so we will have to look into that first.

I've looked into it, and it's not a bug with the allocator, rather a fundamental issue with how the cycle collector handles destruction of cells. It's a variation of the issue with `.acyclic` I mentioned -- I'll create the issue today.

alaviss commented:
> But removing all `.acyclic` annotations is a step back, in that it _hides_ potential issues. If, for example, `ParsedDataNode` forms a cycle, then that's flat out wrong, but it's no longer caught now.

Right now, `{.acyclic.}` is an optimization, not an indicator for correctness. While this PR introduces a process to verify these using sanitizers, I would not consider it reliable on any platforms other than Linux. As far as we know of right now, any of these could have not met the assertion, as they were not tested until now.

Which is why I think we should move ahead slow and steady, reintroduce these annotations with reasoning in code in future PR to contain those changes with a vote of confidence from CI.

Ideally we add a mode in ORC to verify that "acyclic" types don't exhibit cycles so local users can test it easily, but that's work for the future.

Furthermore, as you noted, there are issues with how the runtime handles acyclic as well, so it doesn't inspire confidence that acyclic is safe to apply without proper reasoning and testing.

zerbina commented:
> Right now, `{.acyclic.}` is an optimization, not an indicator for correctness

I somewhat disagree. I see it as an assertion made by the programmer to the compiler that the type isn't really cyclic even though it looks like it. While the compiler has to trust this assertion, although it could - as you said - verify it at runtime, the leak tests act as a strong verification of assumption now. 

> Furthermore, as you noted, there are issues with how the runtime handles acyclic as well

That's not what I said. There are issue with how the runtime handles freeing of cells, but that's not directly related to `.acyclic`.

---

As a compromise, what do you think about the following:
* the `.acylic` pragma is not removed for `CgNode` (refer to https://github.com/nim-works/nimskull/pull/992)
* I fix the cycle collector issue
* I, after this PR is merged, make a PR that reinstates the `.acyclic` pragma in all cases I mentioned

alaviss commented:
> As a compromise, what do you think about the following

Isn't this how I was advocating everything to be done? :P

Sans the CgNode ofc.

But sure, I'll re-instantiate acyclic for that one.

alaviss committed:
cgir: remark CgNode as acyclic

It was marked as such in https://github.com/nim-works/nimskull/pull/992, providing confidence
that it could be correct as well as a paper trail.

saem commented:
> Furthermore, as you noted, there are issues with how the runtime handles acyclic as well, so it doesn't inspire confidence that acyclic is safe to apply without proper reasoning and testing.

Besides `TNode` and friends that are not in fact `acyclic`, the others ones are in terms of their intention. So we can apply `acyclic` based on the behaviour of the implementation or based on intention. I believe @zerbina's perspective is that we should do it by intention and yours is by behaviour.

Going by behaviour has the plus side of having the correct net effect, while taking into account the current implementation's short comings. The downside is that the implementation is a moving target and so is the code that's using it, then we've got two moving targets, plus some tests which may or may not provide a sufficient spec (with failures marked as known issues). Also, since this stuff is lots of threading the needle, one has to actively remember where all the dragons lie, which isn't easy if there isn't directly focused effort on the collector & destructors.

Going by the intention has the plus side of testing the interface as it should be, and there is only one moving target the implementation, while the compiler code and tests together act as a spec (higher likelihood of it being more complete). The downside of course is that we might have some serious issues with the implementation so we _might_ have to compromise on intention correct code, but that's not guaranteed.

I personally favour the intention approach, because that exercises things with and without `acyclic`, and we can note leaving those types as such in this PR. Not to mention, it's easier to remember to take something away, than it is to re-add something. Finally, since working on arc/orc is more indirect work right now based on as the need arises, rather than dedicated effort, the behaviouar based approach is more onerous in terms of having to remember all sorts of edge cases whenever usage gets revisited.

zerbina commented:
> > As a compromise, what do you think about the following
> 
> Isn't this how I was advocating everything to be done? :P

It's not how I understood your suggestion. You didn't mention `CgNode` and you also said "slow and steady", which I consider a single PR to be the opposite of.

saem commented on .github/workflows/leak.yml:
```diff
@@ -0,0 +1,42 @@
+  leaktest:
+    # Skip this for PRs
+    #
+    # FIXME: uncomment this before merge
+    # if: github.event_name != 'pull_request'
```
```suggestion
    if: github.event_name != 'pull_request'
```

Leaving this as a suggestion so it's easier to remember.

saem replied:
```suggestion
    if: github.event_name != 'pull_request'
```

Leaving this as a suggestion so it's easier to remember.

alaviss commented:
> I somewhat disagree. I see it as an assertion made by the programmer to the compiler that the type isn't really cyclic even though it looks like it. While the compiler has to trust this assertion, although it could - as you said - verify it at runtime, the leak tests act as a strong verification of assumption now.

I viewed this one through the lens of the current state of the language to an user, hence usage of the word "right now", as there is no facilities for verifying these assertions, making them more or less an optimization.

The only solution at the moment is to verify via an external provider (LeakSanitizer in this case). If we want this to be treated as a proper assertion, then I believe testing facilities must be provided and documented.

saem commented on compiler/front/main.nim:
```diff
@@ -519,6 +519,17 @@ proc mainCommand*(graph: ModuleGraph) =
   when defined(nimDebugUnreportedErrors):
     addExitProc proc = echoAndResetUnreportedErrors(conf)
 
+  when defined(gcOrc):
```
would disabling gc be better in `nim.nim`, that way it's restricted to the compiler binary, rather than anything that pulls in `main`?

alaviss replied:
Going by the code, this looks more like the main proc to me as you kinda become the `nim` frontend when you use this, but I don't object moving to `nim.nim` if needed.

saem replied:
It's a soft suggestion, just feels symmetric with the change in `nimsuggest.nim`.

saem reviewed (commented):
I'd favour not removing `acyclic` where it actually is valid.

alaviss committed:
ci: disable leak tests for pull requests

This test does not have to run all the time since it is rare that
leaks are introduced to the system.

zerbina commented:
There was a bit of back and forth on this topic now. What's the agreed-on plan now?

If the correct `.acyclic` are removed, then the places I mentioned should at least be marked with a comment to note that `.acyclic` needs to be restored.

alaviss commented:
> If the correct `.acyclic` are removed, then the places I mentioned should at least be marked with a comment to note that `.acyclic` needs to be restored.

sure, I can do that

alaviss committed:
mark types that are likely to be acyclic for restoration

This is a reminder for future zerbina

alaviss commented:
I still think that these annotations should be restored separately from this PR. For two reasons:

- As mentioned before, this PR brings us back to a "correct" (read: leakless) state.
- In the restore PR there would be more space to do things like adding an explainer on why is something acyclic, which I think should be done for all types marked as such.

saem commented:
> I still think that these annotations should be restored separately from this PR. For two reasons:
> 
> * As mentioned before, this PR brings us back to a "correct" (read: leakless) state.
> * In the restore PR there would be more space to do things like adding an explainer on why is something acyclic, which I think should be done for all types marked as such.

If you can run with the above plan then I'm good with it, one less thing for zerbina or I to keep in mind.

zerbina reviewed (approved):
Okay, then I think that's settled.

---

While a task for a later PR, the compiler should also run `GC_fullCollect` in strategic place. Besides leaking the cycles, the root list is also keeps growing, creating another logical memory leak (reintroducing `.acyclic` annotations will mitigate this somewhat).

While I do value speed a bit more, I also value keeping memory usage low during normal compilation.

alaviss commented on nimsuggest/nimsuggest.nim:
```diff
@@ -207,6 +207,16 @@ proc listEpc(): SexpNode =
+  ## Runs a GC collection pass. This procedure will also collect any cycles assuming ORC is in use.
+  ##
+  ## When nimsuggest is being used as a library, no GC mode change will be performed.
+  when isMainModule and defined(gcOrc):
+    GC_enableOrc()
+  GC_fullCollect()
+  when isMainModule and defined(gcOrc):
+    GC_disableOrc()
```
@zerbina do you know if I need to toggle orc state to run the cycle collector?

zerbina replied:
Enabling ORC prior to `GC_fullCollect` is not necessary, but if you want to keep the cycle collector disabled afterwards, the `GC_disableOrc` is needed.

alaviss commented on nimsuggest/nimsuggest.nim:
```diff
@@ -207,6 +207,16 @@ proc listEpc(): SexpNode =
+  ## Runs a GC collection pass. This procedure will also collect any cycles assuming ORC is in use.
+  ##
+  ## When nimsuggest is being used as a library, no GC mode change will be performed.
+  when isMainModule and defined(gcOrc):
+    GC_enableOrc()
```
```suggestion
```

alaviss committed:
nimsuggest: no need to enable orc before collection

The cycle collector runs upon GC_fullCollect(), as such there is no need to enable them.

However running the collector would enable ORC, so we still have to disable it afterwards

alaviss commented:
/merge

I suppose we are all happy with this.

Let's hit the gas then. @zerbina I suppose we will be counting on you for the acyclic restoration.
